### PR TITLE
Transition from Gulp to Webpack

### PR DESCRIPTION
**Pull request by billyhunt**
[Link to ZenHub issue.](https://github.com/usdoj-crt/crt-portal-management/issues/1278)

## What does this change?

Our current build tool, Gulp, is no longer being maintained, and is starting to cause us some problems with active maintenance of the codebase.  This PR is our transition to a more contemporary build tool, Webpack.  Webpack will increase security, add cache-busting to the site so users will not run broken code every deploy, and  allow us to better support the application.

Installation:
Just to be on the safe side, I would wipe all the docker volumes and images, and reinstall.  We want to make sure everything is working as it should.  I would also run an npm install.

Notes: 
* The main thing webpack is doing is rebuilding our Javascript and SCSS assets.  That is where we should focus testing. I have created a checklist as a framework to view everything
* We had some inconsistencies in terms of how we were loading svg's, and it was causing some problems here.  I copied all of them from the US Web Design System into a local folder so they would be easy to access moving forward.
* All of the older javascript I moved into a /components folder.  All necessary files are referenced in a page level bundle.  So, for example, all javascript previously in base.html will now be referenced in base.js, which is imported into base.html through django.webpack.bundler
* Javascript on Dev, Prod, and Staging has a hash in it so we can cache bust.  Locally, there is no hash to simplify hot reloading for development.

## Checklist:
Since this PR touches every page on the site, the testing has to be pretty extensive.  Hopefully this list can be useful in the future. I have grouped the tests by page, with links to click on if you are testing locally.  

https://localhost:8000
  - [x] Open the console, make sure there are no javascript errors
  - [x] Compare the site to https://crt-portal-django-dev.app.cloud.gov/ and make sure all CSS matches
  - [x] Click "Here’s how you know" and make sure dropdown shows dropdown text and closes when clicked again.
  - [x] Make sure text is highlighted in languages menu and menu bar in the header when you mouse over
  - [x] Click "Workplace discrimination or other employment-related problem" and make sure dropdown works.
  - [x] From the same menu, select "Victim of Human Trafficking".  Then select the "Get help from the National Human Trafficking Hotline".  You should see a modal.  Make sure the "Back" and "Continue" buttons work.
  - [x] Check Language dropdown in footer to make sure it works

http://localhost:8000/hate-crime-human-trafficking
  - [x] Open the console, make sure there are no javascript errors
  - [x] Compare the site to https://crt-portal-django-dev.app.cloud.gov/ and make sure all CSS matches
  - [x] Click the "Get help from the National Human Trafficking Hotline" link.  Make sure a modal pops up, and the "Cancel" and "Continue" buttons work. 
  - [x] Make sure the sticky sidebar is sticking, and the correct section is highlighted.

http://localhost:8000/privacy-policy
  - [x] Open the console, make sure there are no javascript errors
  - [x] Compare the site to https://crt-portal-django-dev.app.cloud.gov/ and make sure all CSS matches
  - [x] Make sure the sticky sidebar is sticking, and the correct section is highlighted.

http://localhost:8000/housing-resources
  - [x] Open the console, make sure there are no javascript errors
  - [x] Compare the site to https://crt-portal-django-dev.app.cloud.gov/ and make sure all CSS matches
  - [x] Test all external links.  Make sure a modal appears if the site is directing you to a non Justice Department Site.

https://localhost:8000/report
  - [x] Open the console, make sure there are no javascript errors
  - [x] Compare the site to https://crt-portal-django-dev.app.cloud.gov/ and make sure all CSS matches
  - [x] On first page of report, select Yes or No in the last question, and click "Next".  You should see a modal. Make sure both buttons work.
 - [x] File a report.  Before submitting, edit the results.  Make sure the report works and the edits are in place after submission. 

https://localhost:8000/form/view
  - [x] Open the console, make sure there are no javascript errors
  - [x] Compare the site to https://crt-portal-django-dev.app.cloud.gov/ and make sure all CSS matches
  - [x] Make sure the url has been changed from "https://localhost:8000/form/view" to "http://localhost:8000/form/view/?status=new&status=open&no_status=false"
  - [x] Check to make sure the report you entered previously shows up.
  - [x] Click ALL search filters and make sure the dropdowns work.  Note, some filters like Date have dropdowns within dropdowns.  Make sure those work as well.
  - [x] Do a bulk action, from there check all dropdowns.  Make sure the bulk action works.

http://localhost:8000/form/new/
  - [x] Open the console, make sure there are no javascript errors
  - [x] Compare the site to https://crt-portal-django-dev.app.cloud.gov/ and make sure all CSS matches
  - [x] Click ALL search filters and make sure the dropdowns work.  Note, some filters like Date have dropdowns within dropdowns.  Make sure those work as well.

http://localhost:8000/form/view/1/
  - [x] Open the console, make sure there are no javascript errors
  - [x] Compare the site to https://crt-portal-django-dev.app.cloud.gov/ and make sure all CSS matches
  - [x] Check all dropdowns.  Make sure they open and close, and that you can make selections
  - [x] Pay special attention the the "Assigned to" dropdown.  Select something, then clear it by "x"ing it out.
  - [x] Edit information in the report.  Try both changing the info, and canceling out to make sure it all works.
  - [x] Click the contact complainant button.  Go through all possible options on the modal.
    - [x] Select different forms
    - [x] Print
    - [x] Copy
    - [x] Send

http://localhost:8000/form/dashboard/
  - [x] Open the console, make sure there are no javascript errors
  - [x] Compare the site to https://crt-portal-django-dev.app.cloud.gov/ and make sure all CSS matches
  - [x] Check all dropdowns.  Make sure they open and close, and that you can make selections

### Author

+ [x] If this is a story, run locally and check to make sure all Acceptance Criteria are met. (If any criteria are unclear, ask about them.).
+ [x] Check for, document, and establish a testing plan for any behavior that may vary across environments or is otherwise difficult to test.
+ [x] Check for [accessibility](https://github.com/usdoj-crt/crt-portal/blob/develop/docs/a11y_plan.md).
+ [x] [Tests pass](https://github.com/USDOJ/crt-portal/#tests).

### Reviewer

+ [x] If this is a story, run locally and check to make sure all Acceptance Criteria are met. (If any criteria are unclear, ask about them.).
+ [x] Check for any behavior that may vary across environments or is difficult to test, and ensure that it is well-understood, documented, and that there is a testing plan in place.
+ [ ] Re-check for [accessibility](https://github.com/usdoj-crt/crt-portal/blob/develop/docs/a11y_plan.md).
+ [ ] [Tests pass](https://github.com/USDOJ/crt-portal/#tests).

## Notes for reviewer:

See [PR instructions doc](https://github.com/usdoj/crt-portal/blob/master/docs/pull_requests.md) for full pull request review instructions.

### CODE REVIEW

**Comment by eileen-nava** [comment]
I have a comment about the `localhost:8000/housing-resources` bullet point below:
>  - [ ] Test all external links. Make sure a modal appears if the site is directing you to a non Justice Department Site.

I see a modal when I click some links, but not all.  

Links where I saw modals:
-  Equifax, Experian, and Trans Union.  
- [Homelessness Resources](https://www.hudexchange.info/housing-and-homeless-assistance/)
- [American Bar Association](https://www.americanbar.org/groups/legal_services/flh-home/)
- [Legal Services Corporation (or Legal Aid Offices)](https://www.lsc.gov/find-legal-aid)


Links where I did NOT see modals:
- [Consumer Financial Protection Bureau](https://www.consumerfinance.gov/complaint/).
-  [Fair Housing Accessibility First](https://www.hud.gov/program_offices/fair_housing_equal_opp/accessibility_first_home) 
-  [U.S. Department of Housing and Urban Development](https://www.hud.gov/program_offices/fair_housing_equal_opp/online-complaint).
- [Office of Public & Indian Housing](https://www.hud.gov/program_offices/public_indian_housing/)
- [Office of Inspector General (OIG)](https://www.hudoig.gov/about/where-were-located)
- [Fair Housing Initiatives Program Agencies](https://www.hud.gov/program_offices/fair_housing_equal_opp/partners/FHIP)
- [Fair Housing Assistance Program Agencies](https://www.hud.gov/program_offices/fair_housing_equal_opp/partners/FHAP/agencies)
- [Rent and Utility Payment Assistance](https://www.consumerfinance.gov/coronavirus/mortgage-and-housing-assistance/renter-protections/find-help-with-rent-and-utilities/)
- [State Attorneys General](https://www.usa.gov/state-attorney-general)
- [USA.Gov](https://www.usa.gov/)

**Comment by billyhunt** [comment]
I believe the rule set is if it is a government agency we do not need to show a modal. If it is we do. @ilodidoj can you confirm?

**Comment by ilodidoj** [comment]
@billyhunt @eileen-nava usually for 508 compliance if the URL is pointing to different site the the source origin, modal will be needed.  In our case all federal agencies are part of greater FEDERAL umbrella, therefore no modal. Modal will be needed for non-federal entities as it is now.

**Comment by ilodidoj** [comment]
@billyhunt please check any branch conflict and resolve them before assigning for PR review. CC: @eileen-nava @JackRyan1989

**Comment by eileen-nava** [comment]
> @billyhunt @eileen-nava usually for 508 compliance if the URL is pointing to different site the the source origin, modal will be needed. In our case all federal agencies are part of greater FEDERAL umbrella, therefore no modal. Modal will be needed for non-federal entities as it is now.

I didn't know about this distinction! TIL!  Thanks for sharing.

**Comment by ilodidoj** [comment]
@billyhunt please check the CI/CD error and resolve the issue. crt-django version should be updated as well in package.json; just a thought.  CC: @eileen-nava @JackRyan1989

**Comment by billyhunt** [comment]
Interesting idea on the versioning.  We have not really been doing releases so far.  Do you think we should start @ilodidoj? What should it be?  2.0?  1.1?

File: Dockerfile [hunk 1]
```diff
@@ -37,6 +37,7 @@ RUN \
 
 COPY package*.json /code/
 
+RUN npm install webpack -g
```
Comment (JackRyan1989) [inline Dockerfile:40]: Just noting here that my system seems to hang on this step. We may need to note in the install documentation that if the installation seems to hang up here, users should do an npm install before docker-compose up.

File: crt_portal/cts_forms/templates/forms/pro_template.html [hunk 2]
```diff
@@ -2,6 +2,7 @@
 {% extends "forms/report_base.html" %}
 {% load static %}
```
Comment (JackRyan1989) [inline crt_portal/cts_forms/templates/forms/pro_template.html:3]: I think load static is unnecessary?

File: crt_portal/cts_forms/templates/forms/question_cards/commercial_public_location.html [hunk 3]
```diff
@@ -24,7 +24,7 @@
   {% endif %}
 </fieldset>
 {% block page_js %}
-<script src="{% static 'js/other_show_hide.min.js' %}"></script>
+{% render_bundle 'commercialPublicLocation' %}
```
Comment (JackRyan1989) [inline crt_portal/cts_forms/templates/forms/question_cards/commercial_public_location.html:27]: Does this work without importing the render_bundle from webpack_loader?

File: crt_portal/cts_forms/templates/forms/report_base.html [hunk 4]
```diff
@@ -1,6 +1,7 @@
 {% extends "base.html" %}
 {% load i18n %}
 {% load static %}
```
Comment (JackRyan1989) [inline crt_portal/cts_forms/templates/forms/report_base.html:3]: Can we remove?

File: crt_portal/cts_forms/templates/landing.html [hunk 5]
```diff
@@ -444,6 +445,5 @@ <h3 class="h3__display">{% trans "Need urgent legal help?" %}</h3>
      }
    };
   </script>
```
Comment (JackRyan1989) [inline crt_portal/cts_forms/templates/landing.html:447]: Maybe in a future PR we extract this inline JS to a new component?

File: package.json [hunk 6]
```diff
@@ -39,23 +40,29 @@
...
     "pa11y-ci": "^2.4.2",
-    "prettier": "1.18.2"
+    "prettier": "1.18.2",
+    "sass-loader": "^12.6.0",
+    "style-loader": "^3.3.1",
+    "webpack": "^5.72.0",
+    "webpack-bundle-tracker": "^1.5.0",
+    "webpack-cli": "^4.9.2",
+    "webpack-dev-server": "^4.9.0",
+    "webpack-merge": "^5.8.0"
```
Comment (JackRyan1989) [inline package.json:66]: Do we use webpack merge?

**Review by JackRyan1989 — CHANGES_REQUESTED** [review]
I think the templates can be cleaned up a bit. There are some .html files that only import static files for the JS, and since that is no longer being done by the Django build process, we can remove that import and avoid any confusion there. There's one file that uses a built module without importing webpack_bundler into the template, and I'm not sure if that will cause an issue?

**Review by JackRyan1989 — APPROVED** [review]

**Pull request merged**